### ISSUE TITLE
ONNX Model Predicts different values in CNTK C# than Python Keras Model

### ISSUE THREAD

**Issue by SpicySyntax** [microsoft/CNTK#3608]
I am training a VGG16 based CNN using tranfer learning with Python and Keras. I then convert the model to ONNX format to perform inference from C# but when I test the same image with both models I get different predictions. This makes the predictions from CNTK quite a bit worse for all my images and I am not sure how to troubleshoot where this issue is coming from. Is this an issue with the conversion of the model or am I doing something wrong?

[Here](https://gist.github.com/SpicySyntax/072a035493185df186a61b283b9147b6) I proved the CLI program used for inference with C# (Program.cs) as well as the python script for training and testing the model in Keras (train.py). Both programs test 'crack.jpg' and get different values. (Keras: 0.00033775, CNTK: 0.1526919)

Notes:
- Both images are normalized by dividing the pixel values by 255
- CNTK input vector for the images has dimensions (3, 224, 224, 1, 1, 1)
- Keras input vector has dimensions (1, 224, 224, 3)
- The pixel values seem to be correct in both places:
  - from train.py line 100 `print("{}, {}, {}".format(image[0][0][0], image[0][0][1], image[0][0][2]))` yields: '[0.25882354 0.5372549  0.7411765 ], [0.2627451  0.53333336 0.75686276], [0.2784314 0.5254902 0.8039216]'
  - which seems to match what is expected in Program.cs Lines 126-128 `                Console.WriteLine($"{normalizedCHW[0]}, {normalizedCHW[1]}, {normalizedCHW[2]}");
                Console.WriteLine($"{normalizedCHW[50176]}, {normalizedCHW[50177]}, {normalizedCHW[50178]}");
                Console.WriteLine($"{normalizedCHW[100352]}, {normalizedCHW[100353]}, {normalizedCHW[100354]}");` which yielded: '0.7490196, 0.7647059, 0.7960784
0.5372549, 0.5372549, 0.5294118
0.2588235, 0.2627451, 0.2745098'

**Comment by delzac** [comment]
As you already notice, cntk and keras use different convention (channel first / channel last depending on what backend you used with keras).

Can you also try a np.flip on your channel axis. Cause sometimes, rgb over at one framework might get converted to bgr.

**Comment by SpicySyntax** [comment]
@delzac I have tried alternating the rgb to bgr but it still did not help.
Should the inputs have the same shape in both cases (training and inference)? Should I try reshaping the input in CNTK to match Keras, IE channel last?

**Comment by YanYas** [comment]
This may not fix your issue but just to help debug it, try saving the keras model and the cntk models out and compare them with [Netron](https://github.com/lutzroeder/netron). It should give away the descrepencies between your models pretty quickly. 
I was going to put it down to maybe the intialization values being different but the span of your results is too big for that. I then noticed that you tried saving the model to onnx to infer a result and then I thought 'aha.' The loader (atleast on the C# side that I've actually tested) is a little out of wack. I remember trying it with SqueezeNet and seeing that at certain layers it was outputting the wrong shape, thus always failing.
If you try with the the current release candidate from nuget and  see if that works any better. I certainly hope so.

**Comment by delzac** [comment]
To add on, yes, your input shape must be identical in both cases (training and inference). When onnx saves your model, it will save it as-is. Let us know if you have any more problems!

**Comment by SpicySyntax** [comment]
@YanYas I tried using the 2.7 release candidate but it returns the same value. 

@delzac, in your first message were you suggested that before converting the model to ONNX format I flip the input axis? (Could you show me a code example?) I tried reshaping it from the CNTK C# side and got the following error message `"The leading dimensions of the Value shape '[224 x 224 x 3 x 1 x 1]' do not match the Variable 'Input('input_1_0', [3 x 224 x 224 x 1], [])' shape '[3 x 224 x 224 x 1]'`

**Comment by delzac** [comment]
Now that i looked at this closely again. I think ONNX is saved as channel first and you should not change the model shape. Just use whatever shape onnx demands.

No, no. You do not have to go anything to the saved model. Just attempt to np.flip your image channel axis on the C# side and see if it makes a difference.

There should be no reason why the reason in keras is different than from onnx. If possible, you can download [onnx runtime](https://github.com/Microsoft/onnxruntime) and do inference over there to double check the saved model.

**Comment by SpicySyntax** [comment]
@delzac I took your advice and moved around the Channels.  
If the images is 224 x 224 = WxH let the pixel at a position i,j have the following components for RGB (Rij,Gij, Bij) i and j in [0, 224).

Before hand I collected the floats for the batch in the following format as an List<float> lets call it L:
L[0] = R00, L[1] = R10, L[2]=R20 L[WxH] = G00, L[WxH+1] = G10, L[WxH+2]=G20, L[2xWxH] = B00, L[2xWxH +1] = B10, L[2xWxH +2] B20  and so on... No I updated it so that RGB values for each bixels are contiguous in the list I.E. L[0] = R00, L[1]=G00, L[2]=B00, L[3] = R10, L[4] = G10, L[5]=B00 and so on.

Now the predictions are much closer to the value obtained from Keras (even better in some cases), the example above now gave 0.02853788. I am still confused why I can't seem to get these to match up,
is there anything else I can try before I have to resort to switching to a new framework?

**Comment by delzac** [comment]
The result you get should be identical regardless of what framework you use.

Have you tried every possible ordering of the channel axis? Either rgb, grb, bgr, rbg and so on.

**Comment by SpicySyntax** [comment]
Yes I did, they all resulted in worse results over my tests samples than the simple RGB

**Comment by delzac** [comment]
Hmmm, maybe you might want to take yanyas advice and youse [netron](https://github.com/lutzroeder/netron) to visualise if the model is saved correctly. I think it should be, but doesn't hurt to check.

The next thing to do is to select a few layers and check the weights in keras and onnx to make sure they are the same. You can do that using netron too.

**Comment by SpicySyntax** [comment]
So I viewed them in netron, and it seems when the model was exported to Onnx it somehow held on to the old output before retraining (See images). However since I am using the correct output in C# CNTK side of things this should not be an issue. I could print some weights in Keras and on the C# side of things potentially but I am not sure how to view weights in netron, any suggestions?

**Comment by delzac** [comment]
When you click the layers, a panel will appear. There's a button you can press to inspect the weights

**Comment by SpicySyntax** [comment]
Sorry I forgot to attach these...
ONNX Model:
![vgg16-crack](https://user-images.githubusercontent.com/19228996/54936914-7a73b900-4ef1-11e9-8836-15074ce92880.png)

KERAS Model:
![CrackDetectionVGG16_4](https://user-images.githubusercontent.com/19228996/54936921-7e074000-4ef1-11e9-92fd-f9718b4deb8a.png)

**Comment by SpicySyntax** [comment]
> When you click the layers, a panel will appear. There's a button you can press to inspect the weights
Where at? when I select the first Conv2d I see nothing indicating weight values.
![NN](https://user-images.githubusercontent.com/19228996/54940341-8d3dbc00-4ef8-11e9-9b11-6d9223020f86.PNG)

**Comment by delzac** [comment]
There's a plus sign in the inputs section. Click it.

Also, do you know what backend you used in keras to train the model? Different framework (backend) has slightly different implementation of dropout. In CNTK, dropout during inference is a no-op. I noticed that in the onnx model, dropout is not present.

**Comment by SpicySyntax** [comment]
@delzac I have verified that the weights (although different shapes in the convolutional layers) are the same in both models. I am using Keras with the Tensorflow Backend. After some tweeking I got the C# CNTK backend to predict `0.0006163912` which is much closer to the result obtained from the Keras prediction `0.00033775`. Could the discrepancies be caused by the Keras model performing dropout that is not happening on the C# side? (I thought that Keras only used dropout during the training phase https://stackoverflow.com/questions/47787011/how-to-disable-dropout-while-prediction-in-keras)

**Comment by delzac** [comment]
Hmmm i think it probably has nothing to do with dropout since from the stackoverflow answer it showed that in keras dropout only happens in training.

What were you tweaking that made a difference to the result?

**Comment by SpicySyntax** [comment]
In addition to changing the order of the floats List passed to the CreateBatch method on C# as described above I realized I was flipping the picture 90 Degrees on accident by  flipping the x and y axes. After swapping them back I.E. switching `var rgb = image[i, j];` in the inner loop to `var rgb = image[j, i];` seemed to make the model perform much more closely to the model in Keras.

**Comment by delzac** [comment]
Looking through the netron graphs, the other possible avenue where errors might be introduced is the Flatten layer in keras. The equivalent op in onnx seem to be transpose and then flatten. Does keras Flatten layer really transpose then flatten? You might want to check that out.

**Comment by SpicySyntax** [comment]
It appears Keras does not transpose before flattening because it does not have to:
- The last convolutional layer is 3x3x512x512
- In the ONNX model  the last convolutional layer is 512x512x3x3
  - I 'believe' when this is tranposed is it the aformentioned 3x3x512x512 matrix as in Keras.
-  The Fully connected layers then match up after max pooling

**Comment by delzac** [comment]
Okay got it, onnx transpose image before and after the convolutional stack, thats okay.

After you load the onnx model into cntk, can you save the model out again in native cntk model and inspect it using netron. Check if the representation after loading the onnx model is correct.

Honestly, it seems like all the keras exports to onnx correctly.

**Comment by SpicySyntax** [comment]
I agree, with that.
I am not sure how I can save the model into native CNTK format because we are loading it in as ONNX.
However when I try to save it as ONNX and open the file in netron, no model loads??
Very confused by this, but maybe it is just an issue with the way am using save:
`Function modelFunc = Function.Load(modelFilePath, device, ModelFormat.ONNX);`
 `modelFunc.Save();`

Any ideas?

**Comment by delzac** [comment]
for python, i would write as follows model.save(filepath). Not sure about C# though, sorry

**Comment by SpicySyntax** [comment]
No problem, I am not sure how to save as native CNTK from the C# side. Do you have any suggestions on how I might proceed?

**Comment by delzac** [comment]
@YanYas Might be able to help.
